File: go/algorithms/quick.go
package algorithms

// QuickSort sorts in-place using Lomuto partition quicksort.
func QuickSort(arr []int) {
    if len(arr) == 0 {
        return
    }
    quick(arr, 0, len(arr)-1)
}

func quick(arr []int, lo, hi int) {
    if lo < hi {
        p := partition(arr, lo, hi)
        quick(arr, lo, p-1)
        quick(arr, p+1, hi)
    }
}

func partition(arr []int, lo, hi int) int {
    pivot := arr[hi]
    i := lo
    for j := lo; j < hi; j++ {
        if arr[j] <= pivot {
            arr[i], arr[j] = arr[j], arr[i]
            i++
        }
    }
    arr[i], arr[hi] = arr[hi], arr[i]
    return i
}

File: rust/src/quick.rs
pub fn quick_sort(arr: &mut [i32]) {
    if !arr.is_empty() {
        let len = arr.len();
        quick(arr, 0, (len - 1) as isize);
    }
}

fn quick(arr: &mut [i32], lo: isize, hi: isize) {
    if lo < hi {
        let p = partition(arr, lo, hi);
        quick(arr, lo, p - 1);
        quick(arr, p + 1, hi);
    }
}

fn partition(arr: &mut [i32], lo: isize, hi: isize) -> isize {
    let pivot = arr[hi as usize];
    let mut i = lo;
    for j in lo..hi {
        if arr[j as usize] <= pivot {
            arr.swap(i as usize, j as usize);
            i += 1;
        }
    }
    arr.swap(i as usize, hi as usize);
    i
}
